File: 2022/july/18-07-2022/number-of-submatrices-that-sum-to-target.py
from collections import defaultdict

class Solution:
    def numSubmatrixSumTarget(self, matrix, target: int) -> int:
        # find the rows and columns of the matrix
        n,m = len(matrix) , len(matrix[0])
        # find the prefix sum for each row
        for i in range(n):
            for j in range(1,m):
                matrix[i][j] += matrix[i][j-1]
        ans = 0
        # fix the left boundary of the column
        for start in range(m):
            # fix the right boundary of the column
            for end in range(start,m):
                # a dictionary to map data
                d = defaultdict(lambda:0)
                d[0] = 1
                summ = 0
                # now we do check at each row
                for i in range(n):
                    curr = matrix[i][end]
                    if start > 0:
                        curr -= matrix[i][start-1]
                    summ += curr
                    ans += d[summ - target]
                    d[summ] += 1
        return ans


print(Solution().numSubmatrixSumTarget([[0,1,0],[1,1,1],[0,1,0]], 0))

File: 2022/july/18-07-2022/number-of-submatrices-that-sum-to-target.rs
impl Solution {
  pub fn num_submatrix_sum_target(matrix: Vec<Vec<i32>>, target: i32) -> i32 {
    let n = matrix.len();
    let m = matrix[0].len();

    let mut memo = vec![vec![0;m+1];n+1];
    for i in 0..n {
      for j in 0..m {
        memo[i+1][j+1] = memo[i+1][j] + memo[i][j+1] - memo[i][j] + matrix[i][j];
      }
    }

    let mut result = 0;
    for i in 0..n {
      for j in 0..m {
        for k in 1..=n {
          let ni = i + k;
          if n < ni { break }
          for l in 1..=m {
            let nj = j + l;
            if m < nj { break }

            let v = memo[ni][nj] - memo[ni][j] - memo[i][nj] + memo[i][j];
            if v == target {
              result += 1;
            }
          }
        }
      }
    }
    result
  }
}